integration/container/cdi_test.go

```diff
@@ -0,0 +1,78 @@
+package container // import "github.com/docker/docker/integration/container"
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"testing"
+
+	"github.com/docker/docker/api/types"
+	dclient "github.com/docker/docker/client"
+	"github.com/docker/docker/integration/internal/container"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/docker/testutil/daemon"
+	"github.com/docker/docker/testutil/fakecontext"
+	"gotest.tools/v3/assert"
+	"gotest.tools/v3/skip"
+)
+
+func TestCreateWithCDIDevices(t *testing.T) {
+	skip.If(t, testEnv.OSType != "linux", "CDI devices are only supported on Linux")
+	skip.If(t, !testEnv.DaemonInfo.ExperimentalBuild, "experimental build required for CDI devices")
+	skip.If(t, testEnv.IsRemoteDaemon, "cannot run daemon when remote daemon")
+
+	cdiSpec := `
+cdiVersion: "0.3.0"
+kind: "vendor1.com/device"
+devices:
+- name: foo
+  containerEdits:
+    env:
+    - FOO=injected
+`
+	cdifile := fakecontext.New(
+		t,
+		"/var/run/cdi",
+		fakecontext.WithFile("vendor1-device.yaml", cdiSpec),
+	)
+	defer cdifile.Close()
+
+	var client dclient.APIClient
+	if !testEnv.DaemonInfo.ExperimentalBuild {
+		skip.If(t, testEnv.IsRemoteDaemon, "cannot run daemon when remote daemon")
+
+		d := daemon.New(t, daemon.WithExperimental())
+		d.StartWithBusybox(t)
+		defer d.Stop(t)
+		client = d.NewClientT(t)
+	} else {
+		client = testEnv.APIClient()
+	}
+
+	ctx := context.Background()
+	id := container.Run(ctx, t, client,
+		container.WithCmd("/bin/sh", "-c", "env"),
+		container.WithCDIDevices("vendor1.com/device=foo"),
+	)
+	defer client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{Force: true})
+
+	reader, err := client.ContainerLogs(ctx, id, types.ContainerLogsOptions{
```

Comment:
@corhere I initially tried to implement something that inspects the container here, but did not see the expected environment variable. This is probably because the container config is being based on the unmodified spec, so if the expectation is that the config shows the new envvar, then we would need to do some kind of reconcilliation after making the modifications. How is this currently handled for other entities (e.g. devices or mounts that are requested?).

Reply:
I had been wondering about that. I noticed that when this was raised before (https://github.com/docker/cli/issues/3864#issuecomment-1382393969) it seemed the preference was to indeed have the spec shown by `inspect` be updated to match what was actually run, and that the CDI library has some API to support that.

That said, it seems like this could be solved at the moby level by capturing the OCI spec after all the DeviceRequests are processed and returning _that_ in `docker inspect`; that would _also_ capture the `--gpu` spec mutations which I assume are equally invisible, as I don't see anything in [`daemon/nvidia_linux.go`](https://github.com/moby/moby/blob/2cf63891427d68661bb8e3c58969058c62dbc85d/daemon/nvidia_linux.go) apart from the `updateSpec` hook implementation.

Reply:
Yes, at present the `--gpu` spec mutations are also not visible to the `inspect` command -- although in this case the changes are limited to a single hook as setting two environment variables.

I agree that the goal should be to get these changes visible to daemon, but I would suggest tackling this in a follow-up if possible. Currently, I don't believe that any of the container engines that offer CDI support do this kind of reconciliation -- unliess they use the on-disk `config.json` file as the source of truth when querying a container.

Reply:
The container's OCI spec is based on the container config, not the other way around. The container config is the source of truth. No reconciliation is performed for other entities. I really wish we'd thought to record the maintainers call which you joined where we hashed this all out, because I vaguely recall we concluded reporting the CDI device request as-is when inspecting the container, without back-propagating how it patches the OCI spec, would be acceptable as it is sufficient to flag that that the container's OCI spec has been patched by CDI and therefore any issues reported with that particular container can be assumed to be the responsibility of the CDI spec's vendor until proven otherwise. Reaching a final decision on this aspect of CDI integration is one of the gates for the feature to graduate from experimental @TBBle.

Reply:
Would it be possible to capture the applied CDI spec into the container config or metadata in some way? One of the advantages of CDI is that the specs can be dynamic, e.g. generated just-in-time for that container, and then discarded afterwards, or even updated in-place, I believe.

This could make `docker inspect` very hard to backtrace for troubleshooting, particularly in the context of, say, cri-dockerd which would (I assume) simply translate the CDI field in CRI into DeviceRequests in the Docker API, but which may be referencing a dynamically-generated CDI spec created just-in-time by a kubernetes Device Plugin (or the newer proposed k8s device API whose name I forget, but which is more-suited to just-in-time device allocation), and hence the CDI name is _meaningless_ (probably a GUID or similar) by that point.

Reply:
I was not aware that dynamic CDI specs were a thing. Given that, some mechanism for reporting the applied CDI spec when inspecting a container will be necessary. What affordances does cri-containerd provide for troubleshooting containers which have dynamic CDI specs applied to them?

Reply:
IIRC, in containerd-cri, CDI is applied to the OCI spec before it's passed to the runtime shim, so the config.json on-disk will have the results of applying the CDI annotations. I'm not sure what debugging options are available at the CRI layer though (e.g., can we dump the historical CRI requests that led to this container's state), and I haven't looked into CDI's support for this, beyond the mention in https://github.com/docker/cli/issues/3864#issuecomment-1387030466 that it has APIs to support inspection; so it's possible containerd-cri doesn't currently expose CDI to inspection, just the resulting OCI spec.

Reply:
Thanks for the discussion here @TBBle and for diving into how things are handled in containerd (and similarly in cri-o).   Note that assuming that `docker inspect` returns a `ContainerJSON` struct which includes the `DeviceRequests` in the `HostConfig`, there will be a reference to the requested CDI device names in the `docker inspect` output. This would match what is currently available in other implementations since these also do not read the generated OCI specification for their `inspect` equivalents. This also matches what is currently available for the `--gpus` flag.
```
"DeviceRequests": [
                {
                    "Driver": "",
                    "Count": 4,
                    "DeviceIDs": null,
                    "Capabilities": [
                        [
                            "gpu"
                        ]
                    ],
                    "Options": {}
                }
            ],
```

With regards to:
> Would it be possible to capture the applied CDI spec into the container config or metadata in some way?

One thing to investigate would be to "inject" the requested devices into an empty OCI spec and then store this. This should give a reasonable representation of the edits required for a set of CDI devices. Note that the fesibility of this would have to be properly investigated.

Now from a testing perspective, checking the `HostConfig` doesn't necessarily make sense for this test since this is what we set as input, so I think parsing the logs is fine here.

Reply:
I'd suggest including a HostConfig inspection check in the test, simply because that's a publicly visible API we'll want to maintain, i.e. dockerd-cri's implementation of `critool inspect` will depend on this working, and there's no harm and little cost in making that explicitly tested, since unit tests are amongst other things a form of documentation.